File: ruby/src/detectors/cross-site-request-forgery/cross-site-request-forgery.rb
=begin
 Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 SPDX-License-Identifier: Apache-2.0
=end

# {fact rule=cross-site-request-forgery@v1.0 defects=1}
def cross_site_request_forgery_noncompliant
  user = params[:user_id]
  users_service_domain = params[:users_service_domain]
  # Noncompliant: Complete URL is user-controlled.
  response = Excon.post("#{users_service_domain}/logins", body: {user_id: user}).body
  token = JSON.parse(response)["token"]
end
# {/fact}

# {fact rule=cross-site-request-forgery@v1.0 defects=0}
def cross_site_request_forgery_compliant
  user = params[:user_id]
  users_service_path = params[:users_service_path]
  # Compliant: Only suffix of the URL is controlled by user.
  response = Excon.post("users-service/#{users_service_path}", body: {user_id: user}).body
  token = JSON.parse(response)["token"]
end
# {/fact}

File: jsx/src/detectors/cross_site_request_forgery/cross_site_request_forgery.js
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */


// {fact rule=cross-site-request-forgery@v1.0 defects=1}
var express = require('express')
var app = express()
function crossSiteRequestForgeryNoncompliant()
{
    app.get("/", (req, res) => {
        // Noncompliant: `sameSite` is set to 'none'.
        res.cookie('cookieName', 'cookieValue', { sameSite: 'none', secure: true })
        res.render("index.html")
    })
}
// {/fact}


// {fact rule=cross-site-request-forgery@v1.0 defects=0}
var express = require('express')
var app = express()
function crossSiteRequestForgeryCompliant()
{
    app.get("/", (req, res) => {
        // Compliant: `sameSite` is set to 'lax'.
        res.cookie('cookieName', 'cookieValue', { sameSite: 'lax', secure: true })
        res.render("index.html")
    })
}
// {/fact}